### PR TITLE
Move ServiceAntiAffinityPriority to score plugin

### LINKED ISSUE #86447: Move ServiceAntiAffinityPriority to its Score plugin [Migration Phase 2]
/sig scheduling
/priority important-soon

See #86446 as an example.

/assign @danielqsj

### PR DESCRIPTION
**Pull request by danielqsj**
/sig scheduling
/priority important-soon
/kind feature
/release-note-none

Fix #86447

/assign @ahg-g

### CODE REVIEW

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 1]
```diff
@@ -293,16 +285,104 @@ func (pl *ServiceAffinity) Score(ctx context.Context, state *framework.CycleStat
...
 
 // NormalizeScore invoked after scoring all nodes.
 func (pl *ServiceAffinity) NormalizeScore(ctx context.Context, _ *framework.CycleState, pod *v1.Pod, scores framework.NodeScoreList) *framework.Status {
-	// Note that priorityReduceFunction doesn't use priority metadata, hence passing nil here.
-	err := pl.priorityReduceFunction(pod, nil, pl.sharedLister, scores)
-	return migration.ErrorToFrameworkStatus(err)
+	reduceResult := make([]float64, len(scores))
+	for _, label := range pl.args.AntiAffinityLabelsPreference {
+		if err := pl.updateNodeScoresForLabel(pl.sharedLister, scores, reduceResult, label); err != nil {
+			return migration.ErrorToFrameworkStatus(err)
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]:
```
return framework.NewStatus(framework.Error, err.Error())
```

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 2]
```diff
@@ -293,16 +285,104 @@ func (pl *ServiceAffinity) Score(ctx context.Context, state *framework.CycleStat
...
+		scores[i].Score = int64(nodeScore)
+	}
+	return nil
+}
+
+// updateNodeScoresForLabel updates the node scores for a single label. Note it does not update the
+// original result from the map phase directly, but instead updates the reduceResult, which is used
+// to update the original result finally. This makes sure that each call to updateNodeScoresForLabel
+// receives the same mapResult to work with.
+// Why are doing this? This is a workaround for the migration from priorities to score plugins.
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: we should have a TODO here

**Review by draveness — CHANGES_REQUESTED** [review]

**Comment by danielqsj** [comment]
@draveness thanks, PR updated. Please take another look

**Review by draveness — COMMENTED** [review]
Thanks

/lgtm

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 3]
```diff
@@ -293,16 +285,105 @@ func (pl *ServiceAffinity) Score(ctx context.Context, state *framework.CycleStat
...
+// to a single plugin.
+// TODO: This will be deprecated soon.
+func (pl *ServiceAffinity) updateNodeScoresForLabel(sharedLister schedulerlisters.SharedLister, mapResult framework.NodeScoreList, reduceResult []float64, label string) error {
+	var numServicePods int64
+	var labelValue string
+	podCounts := map[string]int64{}
+	labelNodesStatus := map[string]string{}
+	maxPriorityFloat64 := float64(framework.MaxNodeScore)
+
+	for _, hostPriority := range mapResult {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: use ```node``` instead of ```host``` here and all places below

**Review by ahg-g — COMMENTED** [review]
one small nit

**Comment by ahg-g** [comment]
revert changes to pkg/scheduler/algorithm_factory.go and rebase.

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 4]
```diff
@@ -293,16 +285,105 @@ func (pl *ServiceAffinity) Score(ctx context.Context, state *framework.CycleStat
...
+
+	// Pods matched namespace,selector on current node.
+	var selector labels.Selector
+	if services, err := schedulerlisters.GetPodServices(pl.serviceLister, pod); err == nil && len(services) > 0 {
+		selector = labels.SelectorFromSet(services[0].Spec.Selector)
+	} else {
+		selector = labels.NewSelector()
+	}
+
+	if nodeInfo.Pods() == nil || len(nodeInfo.Pods()) == 0 || selector.Empty() {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: remove ```nodeInfo.Pods() == nil ||```

Comment (danielqsj) [reply pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: done

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go (hunk 3 above)
Comment (danielqsj) [reply pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: done

**Comment by danielqsj** [comment]
@ahg-g  thanks. Comments addressed. Please take another look. :)

**Pull request merged**